libraries/tlv-account-resolution/src/account.rs

```diff
@@ -66,6 +68,48 @@ where
     Ok(Pubkey::find_program_address(&pda_seeds, program_id).0)
 }
 
+/// Resolve a pubkey from key data configuration.
+fn resolve_key_data<'a, F>(
+    key_data: &KeyData,
+    instruction_data: &[u8],
+    get_account_key_data_fn: F,
+) -> Result<Pubkey, ProgramError>
+where
+    F: Fn(usize) -> Option<(&'a Pubkey, Option<&'a [u8]>)>,
+{
+    match key_data {
+        KeyData::Uninitialized => Err(ProgramError::InvalidAccountData),
+        KeyData::InstructionData { index } => {
+            let key_start = *index as usize;
+            let key_end = key_start + PUBKEY_BYTES;
+            if key_end > instruction_data.len() {
+                return Err(AccountResolutionError::InstructionDataTooSmall.into());
+            }
+            Ok(Pubkey::new_from_array(
+                instruction_data[key_start..key_end].try_into().unwrap(),
+            ))
+        }
+        KeyData::AccountData {
+            account_index,
+            data_index,
+        } => {
+            let account_index = *account_index as usize;
+            let account_data = get_account_key_data_fn(account_index)
+                .ok_or::<ProgramError>(AccountResolutionError::AccountNotFound.into())?
+                .1
+                .ok_or::<ProgramError>(AccountResolutionError::AccountDataNotFound.into())?;
+            let arg_start = *data_index as usize;
+            let arg_end = arg_start + PUBKEY_BYTES;
+            if account_data.len() < arg_end {
+                return Err(AccountResolutionError::AccountDataTooSmall.into());
+            }
+            Ok(Pubkey::new_from_array(
+                account_data[arg_start..arg_end].try_into().unwrap(),
+            ))
+        }
+    }
+}
+
 /// `Pod` type for defining a required account in a validation account.
 ///
 /// This can either be a standard `AccountMeta` or a PDA.
```

Comment:
I can't comment further down, but can you specify that `discriminator` can also refer to dynamic pubkey data on line 120-122?

Reply:
Ahh yep I missed this documentation.